packages/@aws-cdk/core/lib/validation/private/report.ts

```diff
@@ -0,0 +1,190 @@
+import * as os from 'os';
+import { table } from 'table';
+import { ConstructTrace, ConstructTree } from './construct-tree';
+import { Node } from '../../private/tree-metadata';
+import * as report from '../report';
+/**
+ * The report emitted by the plugin after evaluation.
+ */
+export class ValidationReportFormatter {
+  constructor(private readonly tree: ConstructTree) {}
+
+  /**
+   * Get the stack trace from the construct node metadata.
+   * The stack trace only gets recorded if the node is a `CfnResource`,
+   * but the stack trace will have entries for all types of parent construct
+   * scopes
+   */
+  private getTraceMetadata(node?: Node): string[] {
+    if (node) {
+      const construct = this.tree.getConstructByPath(node.path);
+      if (construct) {
+        const trace = construct.node.defaultChild?.node.metadata.find(meta => !!meta.trace)?.trace ?? [];
+        return Object.create(trace);
+      }
+    }
+    return [];
+  }
+
+  /**
+   * Construct the stack trace of constructs. This will start with the
+   * resource that has a violation and then go up through it's parents
+   */
+  private getConstructTrace(node: Node, locations?: string[]): ConstructTrace {
+    const trace = this.tree.getTraceFromCache(node.path);
+    if (trace) {
+      return trace;
+    }
+    const metadata = locations ?? this.getTraceMetadata(node);
+    const thisLocation = metadata.shift();
```

Comment:
We need a better way of handling this. If we reverse the tree then this won't work. Also I'm not sure how this will work with JSII languages.